### PR TITLE
Hashing, caching and fast-forwarding

### LINKED ISSUE #119: Hashing of nodes
Originally reported by: **Andrea Cepellotti (Bitbucket: [acepellotti](https://bitbucket.org/acepellotti), GitHub: [cepellotti](https://github.com/cepellotti))**

----------------------------------------

Implement a node hashing. In this way, it would be possible to immediately recognise whether a node has been already stored in the DB, or if a calculation has been run already, which would return a result immediately.
Note also, this would massively speed up the debugging of workflows (repeating many times the same things): we might immediately understand (create a method for it) whether a wf step has calculations to be executed or not.

----------------------------------------
- Bitbucket: https://bitbucket.org/aiida_team/aiida_core/issue/119

### PR DESCRIPTION
**Pull request by greschd**
Fixes #119.

**Edit / Note:** The description below is somewhat out of date, as it describes the initial state of the PR. For a current description, either follow the changes proposed / discussed subsequently, or refer to the descriptions in the documentation made in this PR.

The hashing, caching and fast-forwarding project is a work in progress. Since it is nearing completion, I'm creating this PR to start collecting feedback.

### Hashing

The ``hashing.make_hash`` creates a hash from a given Python datastructure. Nested structures, are handled by recursive hashing, meaning that for example a dict is turned into 
```
{key: make_hash(value) for key, value in dict.items()}
```
before hashing it. AiiDA ``Folder`` data can also be hashed, by recursively taking the hash of filenames and contents. For floating point numbers, the last 4 bits of the mantissa are truncated before creating the hash. However, this does not apply to ``ArrayData``, where the array is saved as a file on disk and therefore handled by the ``Folder`` hashing.

Using the ``make_hash`` function, the ``Node`` has a ``get_hash`` method. For basic nodes, this method creates a hash from ``get_attrs()``, the database folder, and the ``__version__`` of the module where the ``Node`` class is defined.  If there are errors, ``get_hash`` returns ``None``.

``Node`` subclasses can define a class-level ``_hash_ignored_attributes`` list, with names of attributes that will not be taken into account when creating the hash. **Edit:** ``_updatabale_attributes`` are also ignored now.

The ``WorkCalculation`` subclass adds ``get_inputs()`` to the objects that are hashed in ``get_hash``, and exposes the ``_hash_ignored_inputs`` to ignore certain inputs.

### Caching

When a ``Node`` is stored, the hash is saved as an extra. When subsequent nodes are stored, caching can be enabled by setting ``use_cache=True`` in the ``store`` method. This means that a node of the same type and hash will be returned if it exists. Otherwise a new node is created.

Nodes can determine whether they can be used as cache by implementing ``_is_valid_cache``. This is used to avoid using failed calculations as cache.

The ``verdi rehash`` command can be used to re-calculate the hash either for all nodes, or just for a specific class.

### Fast-forwarding

Fast-forwarding is implemented by using caching for the calculation of a given ``Process``. If fast-forwarding is enabled and there is already such a calculation, this finished calculation will be used. The subsequent steps of running the process are then skipped, and it is done immediately.

The process decides whether fast-forwarding is used based on the ``_fast_forward_enabled`` method. There are two option to set this: 

* First priority: Passing the ``_fast_forward=True / False`` input to the process.
* Second priority: A config file ``cache_config.yml`` in the ``.aiida`` folder. A default can be set, and fast-forwarding for specifc calculation / process classes can be enabled or disabled (by class name). An example file would look like this:

```
my_profile_name:
  use_cache:  # could be omitted (it's the default)
    default: False
  fast_forward:
    default: False  # could be omitted (it's the default)
    enabled: 
      - SomeCalculation
        OtherCalculation
        SomeWorkChainClass
```

Since the fast-forwarding is implemented at the level of ``Process``, it doesn't work for ``InlineCalculation``, and ``JobCalculation`` that is not launched via ``work.run.run`` or ``work.run.submit``. 

### To-Do's

**EDIT:** Updated To-Do's after a discussion with Giovanni.

- [x] Also ignore ``_updatable_attributes`` in the hash, in addition to ``_hash_ignored_attributes``.
- [x] Check if only ``WorkCalculation`` adds the inputs to the hash, or if ``JobCalculation`` also does it. If it doesn't, this needs to be fixed. **Update:** This is implemented in ``_get_objects_to_hash`` of the ``AbstractCalculation``, meaning that it works for both ``WorkCalculation`` and ``JobCalculation``.
- [x] Check that ``plum`` is compatible with the caching, and change ``.travis.yml`` to use a release version. **Update:** Plum version "0.7.11" (not merged) is compatible, but the goal is to merge with plum >= 0.9.
- [x] ``verdi rehash`` should take PKs as variadic arguments. Add a ``-a/--all`` option to rehash all nodes.
- [x] Avoid ambiguitiy in cache_config.yml, by using the full type string for cached classes
- [x] Document that ``_is_valid_cache`` can be used to permanently disable fast-forwarding on a calculation class. Also add some tests using this, e.g. to have a calculation which creates random output. **Update:** To disable caching on a class level, the ``_cacheable`` attribute was added. This avoids doing a hash query when the class cannot be cached at all. This does _not_ mean that the nodes don't have a hash, though.
- [x] Copy nodes instead of just returning them when there is a cache hit. This means that caching will be useless in terms of memory use, but allows for the fast-forwarding functionality and retains the integrity of the graph. Nodes that are copied should get a ``cached_from`` extra, as a stand-in for the ``LinkType.CACHE`` we might want to add if lazy / reference copying is implemented.
- [x] Document what ``ignored_folder_content`` (in ``make_hash``) does. **Update:** This is exposed to the ``get_hash()`` method, and is used to ignore the ``raw_input`` for the ``JobCalculation``. 
- [x] ~~For Folder hashing, handle symbolic links separately. The link target should be hashed instead of the file content, with ``ps`` as type string / salt.~~
- [x] Add documentation
- [x] Make ``_get_config()`` configurable through a keyword argument giving the file path.
- [x] Add tests for caching configuration.
- [x] Make setting the caching default work, or remove it completely.
- [x] Merge the notions of `use_cache` and `fast_forward`.
- [x] Fix and add tests for caching ``JobCalculation`` instances.
- [x] Add "advanced" section to the documentation.
- [x] Add ``has_failed`` and ``has_finished`` methods to AbstractCalculation, implement in InlineCalculation.
- [x] Take Computer uuid into account when creating the Node hash.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -51,6 +51,8 @@ install:
     - pip install -U pip wheel setuptools
     # Install AiiDA with some optional dependencies
     - pip install .[REST,docs,atomic_tools,testing]
+    # Unreleased plum version required for fast-forwarding
```
Comment (greschd) [inline .travis.yml]: This should be removed before merging, after creating a patched ``plum`` 0.7.x release.

**Comment by greschd** [comment]
@sphuber Why was this closed? I am working on the To-Dos above.

**Comment by greschd** [comment]
@giovannipizzi We discussed handling symlinks separately for the Folder hash. However, the ``aiida.common.folder.Folder`` API doesn't seem to have any way to distinguish regular files from symlinks. Is that planned, or should we leave the current behavior (just reading the file content) for now?

**Comment by greschd** [comment]
We could just expose ``os.path.islink`` through the ``Folder`` -- however I don't know if this translates easily to the planned object store interface.

**Comment by giovannipizzi** [comment]
Good point. I remember we discussed symlinks for the new object store interface with @sphuber some time ago, but I don't remember what we decided. Thinking better to it, I think it's ok to keep the interface as it is - as long as we hash repository folders and not remote folders. I don't think it should happen that a repository folder has symlinks, in the current implementation. However, maybe keep a comment in the docstring of the hashing function about this, to avoid potential issues in the future.

**Comment by greschd** [comment]
Okay, I will do that :+1:

**Comment by greschd** [comment]
@giovannipizzi Another question: For the class names in the config file, we decided to use the fully qualified name instead of just the ``__name__`` to avoid accidental matches. I think a good way to implement this would be to actually load these classes when building the configuration. The advantages of this would be
* If the class is defined in some hidden submodule and then exposed via star-imports, the more common (exposed) name can be used.
* If a class doesn't exist, or there's a typo, the user gets immediate feedback.

Now the problem with this approach is that it needs to load the classes, which might need the dbenv. The question is whether it's acceptable to do this when the caching configuration is loaded. Should I
1. Keep it as is, and just put the tests for the caching configuration into db tests?
2. Not load the dbenv in the configuration function, and accept that an error will be raised if the caching configuration happens before the ``load_dbenv``?
3. Try to work around loading the classes altogether?

**Comment by greschd** [comment]
After considering how to re-create the links of cached calculations and workchains, I think that fast-forwarding is only really possible for calculations, not workchains.

For calculations, the case is simple: The calculation which uses the cache keeps its own input links. For the outputs, it creates a copy of all the outputs of the cached calculation. The new calculation is the unique creator of this output data.

Workchains on the other hand can return data, making this pretty much impossible. For example, let's say we have a "echo" workchain W which just returns its input. Given input D1, it will return D1.
Now if the workchain is called with D2 which is identical to D1, it should obviously return D2. But it's not obvious how to identify that D2 is the input that needs to be returned. It's also not possible to distinguish whether D1 was really just returned, or if it was the result of some query. An even worse example is if W returns one of the parents of D1 instead of D1 itself.

So in conclusion, workchains can make links that go against the direction of the DAG. Because it's not clear how those links are determined without actually running the workchain, it's impossible to re-create them when trying to fast-forward a workchain.

**Comment by greschd** [comment]
@giovannipizzi @sphuber I'm done with this, except for the small incompatibility with ``plum``. The only thing that needs to be changed in ``plum`` is that it doesn't fail if one tries to run an already ``FINISHED`` calculation.

I think it makes sense for you look at this now. We can then decide if we want to wait for the new plum version, or just create a bugfix release from https://github.com/aiidateam/plumpy/pull/7.

**Comment by greschd** [comment]
Ready for review, I finished the things we discussed via Skype.

**Comment by greschd** [comment]
Ready for review, I finished the things we discussed on Skype.

About the tests: It seems they fail with some degree of randomness (sometimes it fails, sometimes it doesn't). What happens is some of the calculations in the daemon test are ``SUBMISSIONFAILED``. I can't say with certainty that it doesn't have to do with caching, but it does seem unlikely.

**Comment by sphuber** [comment]
I recall this happening before as well, but I cannot say for sure from your example that it is the exact same thing. What used to happen that sometimes the ssh connection had issues causing the submissions to fail. A simple restart of the test job would often fix it. However, I remember the message of the ssh problem to be explicitly visible, but in your example I could not find it. This might however just be because there have been some changes in the daemon that changed the logging. For example in the current daemon logging, the reports of `WorkChain`s are no longer recorded. This is an outstanding issue.

**Comment by greschd** [comment]
Ok, I think it's safe to say that's a separate issue then.

The reason the log isn't shown there is because I was checking the cached calculations before the regular ones. They also fail because the calculation they should cache from failed. This (the order of checking) is fixed in my last commit.

If I had to take a wild guess I'd say this issue is either a misconfiguration of the torquessh / daemon setup, a glitch in the Travis infrastructure, or a combination of these.

File: aiida/common/caching.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        return func(*args, **kwargs)
+    return inner
+
+@_with_config
+def get_use_cache(node_class=None):
+    if node_class is not None:
+        enabled = node_class in _CONFIG[config_keys.enabled]
+        disabled = node_class in _CONFIG[config_keys.disabled]
+        if enabled and disabled:
+            raise ValueError('Invalid configuration: Fast-forwarding for {} is both enabled and disabled.'.format(node_class))
```
Comment (sphuber) [inline aiida/common/caching.py]: If I understand correctly, we merged the notion of `use_cache` and `fast_forward` to just using `use_cache`. Maybe it is better to use `using cache` insread of `Fast-forwarding` in the exception message.

File: docs/source/caching/index.rst [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+Additionally, there are two methods you can use to disable caching for particular nodes:
+
+* The :meth:`._is_valid_cache` method determines whether a particular node can be used as a cache. This is used for example to disable caching from failed calculations.
+* Node classes have a ``_cacheable`` attribute, which can be set to ``False`` to completely switch off caching for nodes of that class. This avoids performing queries for the hash altogether.
+
+There are two ways in which the hash match can go wrong: False negatives, where two nodes should have the same hash but do not, or false positives, where two different nodes have the same hash. It is important to understand that false negatives are **highly preferrable**, because they only increase the runtime of your calculations, as if caching was disabled. False positives however can break the logic of your calculations. Be mindful of this when modifying the caching behaviour of your calculation and data classes.
+
+.. _caching_provenance:
+
```
Comment (giovannipizzi) [inline docs/source/caching/index.rst]: I would add something like the following section
```rst
What to do if caching is used but you don't want to
---------------------------------------------------
While this should not happen, there are a few edge cases where this could. One example is when one uses a parser that lives in a different python module than the class itself (maybe even set via `calc.set_parserclass`) or maybe one forgets to increase the version number of the python package (that is used inside the hash).

In general, the "correct" solution to this problem is to properly set a version number (and for calculation plugins that depend on external parsers, depend on an exact version of them, and increase your version number when you start depending on a newer version of the parser).

However, in development, this could happen. In this case, if you see a calculation that is not running as you would like (and using results from an older calculation, but you want it to run again):

1. Check if it has a `_aiida_cached_from` attribute. If it has not, then it is not a problem of caching.
2. Go to that calculation (using e.g. `cached_from_node = load_node(uuid)`), and call the method `cached_from_node.clear_hash()` (optional, as this is fixed by the next step).
3. Check if there are more nodes with the same hash: 
   `equivalent = Node.get_all_same_nodes(node.get_hash())`, and for each of them clear the hash::
  
  for node in equivalent:
      node.clear_hash()

4. run your calculation again: this time it should be run again

If you instead think it is a bug in the AiiDA implementation, open an issue (with enough information to be able to reproduce the error, otherwise it is hard for us to help you) in the AiiDA GitHub repository [link here].
```

Note: here I'm assuming that there is (otherwise we should add it) a `Node.clear_hash()`, probably just deleting the `hash` attribute.
I'm also assuming the existence of a `Node.get_all_same_nodes(hash)` (feel free to change the name) to get not any equivalent node, but a list of all nodes with a given hash (always a list, possibly empty - it's a simple query, but better to provide a method).

What do you think of this proposal? I think the usecase described at the beginning of the docs I wrote here explains a case where such a functionality would be useful.

File: aiida/backends/tests/work/workChain.py [hunk 4]
```diff
@@ -385,6 +386,56 @@ def _run_with_checkpoints(self, wf_class, inputs=None):
...
+                raise ValueError
+
+        self.wf_class_broken = ReturnInputsFastForward
+
+    def tearDown(self):
+        super(TestFastForwardingWorkChain, self).tearDown()
+        super(TestFastForwardingWorkChain, self).tearDownClass()
+        super(TestFastForwardingWorkChain, self).setUpClass()
+
+    def test_hash(self):
```
Comment (sphuber) [inline aiida/backends/tests/work/workChain.py]: I was a bit confused by this test as I remembered the discussion that `WorkCalculations` can not be cached, but then here there is actually a test that checks that the caching of `WorkChains` works by passing `_use_cache=True` as an input. Granted, it only checks whether the resulting hashes are equal, but I wondered whether the second workchain was actually taken from the cache or not.

After a lot of digging, I noticed that caching is disabled for `WorkCalculations` because the `_cacheable` property is set to `False` for the `AbstractCalculation` and then reenabled only for the `InlineCalculation` and `JobCalculation`. Additionally I found that you really enforce this behavior, by putting an extra safeguard in the `_store_from_cache` method of the `Node` class in `aiida.orm.implementation.general.node`, by checking whether the node has any `RETURN` links, in which case you don't use the cache, just in case people subclass and override the `_cacheable` property.

Ultimately, I think the current implementation results in the correct behavior, however, I think it would be good to describe this implementation detail in the developer's guide somewhere. It is too specialized for users, but I think it is very important knowledge for developers and it was a bit tricky to track down.

File: aiida/orm/implementation/general/calculation/job/__init__.py [hunk 5]
```diff
@@ -37,6 +36,29 @@ class AbstractJobCalculation(object):
...
 
+    _cacheable = True
+
+    @classproperty
+    def _hash_ignored_attributes(cls):
+        # _updatable_attributes are ignored automatically.
+        return super(AbstractJobCalculation, cls)._hash_ignored_attributes + [
+            'queue_name',
+            'max_wallclock_seconds',
+            'retrieve_temporary_list',
```
Comment (giovannipizzi) [inline aiida/orm/implementation/general/calculation/job/__init__.py]: In terms of reducing false positives, wouldn't it be better to also hash `retrieve_temporary_list`? If this list changes, the parser might fail (or return less/more output nodes), so changing the behaviour - and in this case I would say we don't want to cache it. Or more generally, when can the retrieve_temporary_list change without the plugin version change? (Or, if the plugin allows the user to change it, it will have to appear in some input, so it will be hashed). Maybe I'm missing a use case here?

File: aiida/orm/implementation/general/calculation/job/__init__.py [hunk 6]
```diff
@@ -37,6 +36,29 @@ class AbstractJobCalculation(object):
...
     """
 
+    _cacheable = True
+
+    @classproperty
+    def _hash_ignored_attributes(cls):
+        # _updatable_attributes are ignored automatically.
+        return super(AbstractJobCalculation, cls)._hash_ignored_attributes + [
+            'queue_name',
+            'max_wallclock_seconds',
```
Comment (giovannipizzi) [inline aiida/orm/implementation/general/calculation/job/__init__.py]: For `queue_name` and `max_wallclock_seconds`, while I see the reason for not caching, is this really something we want to do? If yes, then for consistency we should also ignore the other things that are setting properties on the scheduler? 
```
        custom_sched_commands = self.get_custom_scheduler_commands()
        if custom_sched_commands:
            job_tmpl.custom_scheduler_commands = custom_sched_commands

        job_tmpl.import_sys_environment = self.get_import_sys_environment()

        job_tmpl.job_environment = self.get_environment_variables()

        queue_name = self.get_queue_name()
        if queue_name is not None:
            job_tmpl.queue_name = queue_name
        priority = self.get_priority()
        if priority is not None:
            job_tmpl.priority = priority
        max_memory_kb = self.get_max_memory_kb()
        if max_memory_kb is not None:
            job_tmpl.max_memory_kb = max_memory_kb
        max_wallclock_seconds = self.get_max_wallclock_seconds()
        if max_wallclock_seconds is not None:
            job_tmpl.max_wallclock_seconds = max_wallclock_seconds
        max_memory_kb = self.get_max_memory_kb()
        if max_memory_kb is not None:
            job_tmpl.max_memory_kb = max_memory_kb
```
(this comes from the ```JobCalculation._presubmit()``` method).
On the other hand, some of these flags might change the actual behaviour of the code (crash if not enough memory, run with different options if environment variables are different, ...). What is a good compromise? 

In any case, in the docs this choice is not documented, so maybe the info on what is used for the various classes (like: data -> everything; JobCalcs: this above; InlineCalcs: inputs; other Calcs: not cached; ...) should be explained to the user. (and maybe for a subclass how to change this - maybe for a plugin you might want to add more?)

File: docs/source/caching/index.rst [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    profile-name:
+      default: False
+      enabled:
+        - aiida.orm.calculation.job.simpleplugins.templatereplacer.TemplatereplacerCalculation
+        - aiida.orm.data.base.Str
+      disabled:
+        - aiida.orm.data.base.Float
+
+This means that caching is enabled for ``TemplatereplacerCalculation`` and ``Str``, and disabled for all other classes. In this example, manually disabling ``aiida.orm.data.base.Float`` is actually not needed, since the default value for all classes is already ``False``. Note also that the fully qualified class import name (e.g., ``aiida.orm.data.base.Str``) must be given, not just the class name (``Str``). This is to avoid accidentally matching classes with the same name.
```
Comment (sphuber) [inline docs/source/caching/index.rst]: The following line: "since the default value for all classes is already ``False``" may be a bit confusing. By default all `Node` classes are in fact marked as cacheable, in terms of implementation. Only `Calculation` nodes that are not `InlineCalculation` or `JobCalculation` have `_cacheable = False`. But in this line I think you refer to the fact that in this particular config example, the default for caching globally is set to `False`. However, with this explanation a user might think that if he sets the default to `True` in the config, caching is still going to be disabled for the `Float` class, when in reality it is not.

Maybe we can write that in general all node class are cacheable, except if the default in the config is set to `False` or the class is specifically disabled. With the additional exception that `WorkCalculations` and their derivatives are always disabled and cannot be enabled through the configuration.

**Comment by greschd** [comment]
Thanks for the review! I'll go through the changes soon -- doesn't look like there are any major things to do.

File: aiida/orm/implementation/general/node.py [hunk 8]
```diff
@@ -1604,6 +1653,64 @@ def __del__(self):
...
+            },
+            self.folder,
+            computer.uuid if computer is not None else None
+        ]
+
+    def rehash(self):
+        self.set_extra('hash', self.get_hash())
+
+    def get_same_node(self):
+        if not self._cacheable:
```
[the remark this answers is not in the document]
Comment (greschd) [reply aiida/orm/implementation/general/node.py]: Yes, that's exactly what it does. I added a docstring, and since this function is only used internally, I changed the name with a leading underscore.

Also, I moved the implementation to ``_iter_all_same_nodes`` which returns an iterator over all same nodes. The advantage is that I can re-use the same code for the ``get_all_same_nodes`` you propose below. I didn't use ``get_all_same_nodes`` directly because then I would have to evaluate the full query (``.all()`` instead of ``.iterall()``) also for the case when only one match is needed.

File: aiida/orm/implementation/general/node.py [hunk 9]
```diff
@@ -1128,9 +1139,12 @@ def iterextras(self):
             # added (in particular, we do not even have an ID to use!)
             # Return without value, meaning that this is an empty generator
             return
-            yield  # Needed after return to convert it to a generator
-        for _ in self._db_iterextras():
-            yield _
+            yield # Needed after return to convert it to a generator
+        for extra in self._db_iterextras():
+            # Don't return if key == hash
+            if extra[0] == 'hash':
```
[the remark this answers is not in the document]
Comment (greschd) [reply aiida/orm/implementation/general/node.py]: Totally agree. I had to change some of the tests which check setting extras because they didn't expect the automatically-generated ``_aiida_hash``, but other than that "un-hiding" the extra worked fine.

Where in the documentation should we put that ``.. note::`` box?

File: aiida/common/caching.py (hunk 2 above)
Comment (greschd) [reply aiida/common/caching.py]: Nice catch, thanks!

File: aiida/orm/implementation/general/calculation/job/__init__.py (hunk 6 above)
Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: As long as the only possible change is that a job can fail if the setting is wrong (e.g. time limit too low) there shouldn't be a problem, because failed calculations are anyway not taken into account when using the cache. Environment variables are a different story because they might change the result, but there's also a grey area, where a calculation _could_ behave differently if e.g. the memory limit is different, but that doesn't strike me as very realistic.

So maybe it's ok to add ``'max_memory_kb'`` and ``'priority'`` to the ignore list? Or would you rather go the conservative way of not ignoring any of them?

Obviously this should be noted in the documentation.

File: aiida/orm/implementation/general/calculation/job/__init__.py (hunk 5 above)
Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: Absolutely, I think I somehow misunderstood what ``retrieve_temporary_list`` does.

File: aiida/orm/implementation/general/calculation/job/__init__.py (hunk 6 above)
Comment (giovannipizzi) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: If we keep the other things like the queue, then ok to add also max_memory_kb and priority I think.

File: docs/source/caching/index.rst (hunk 3 above)
Comment (greschd) [reply docs/source/caching/index.rst]: I added this section now, and added the two methods.

The ``get_all_same_nodes`` is implemented as an instance method, not a classmethod. This is because the query takes into account the Node class also, not only its hash.

File: aiida/orm/implementation/general/calculation/job/__init__.py (hunk 6 above)
Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: Ok, I added them now. However, I'm now having doubts whether it's a good idea to put this in the documentation. My concern is that when this is inevitably changed in the future, the documentation will most probably not reflect that change. I don't think there's a simple way to keep it up-to-date.

What do you think? Should I put it in anyway?

File: docs/source/caching/index.rst [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
```
[the remark this answers is not in the document]
Comment (greschd) [reply docs/source/caching/index.rst]: Unless the metaclass of the Node has a custom ``__str__``, yes. A more stable way to get it would be ``Str.__module__ + '.' + Str.__name__``.

I now added a snippet on how to get the full name to the documentation.

File: aiida/backends/tests/work/workChain.py (hunk 4 above)
Comment (greschd) [reply aiida/backends/tests/work/workChain.py]: I probably forgot to remove that test after deciding that caching doesn't work for ``WorkCalculations`` -- I deleted it now.

As for ``_cacheable`` and the ``RETURN`` links, I added a small section in the developer's guide.

File: docs/source/caching/index.rst (hunk 7 above)
Comment (greschd) [reply docs/source/caching/index.rst]: I changed the wording a bit to hopefully make it clear that it's referring to the default configuration value. As for the ``_cacheable`` property, I don't think we need to tell the user about this at that point. The fact that workchains and workfunctions cannot be cached is mentioned in the introductory section.

The user should need to know about ``_cacheable`` only if he wants to disable it manually for his classes, which is what the "how are cached nodes matched" section is for.

**Comment by greschd** [comment]
My pleasure. Thanks for all your help!

File: aiida/orm/implementation/general/calculation/job/__init__.py (hunk 5 above)
Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: @giovannipizzi After actually running this code on my project, I think ``retrieve_temporary_list`` should be an ``_updatable_attributes``, and therefore ignored. When I launch a calculation without ``retrieve_temporary_list``, it changes to an empty list _after_ the calculation is stored. This means that nodes will not get their original hash when you re-hash them, breaking caching in this case.

Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: Since the other "retrieve" attributes are also updatable, I think putting it there is the correct way to solve this:

```
        self._updatable_attributes = (
            'state', 'job_id', 'scheduler_state',
            'scheduler_lastchecktime',
            'last_jobinfo', 'remote_workdir', 'retrieve_list',
            'retrieve_singlefile_list'
        )
```

Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: I'll need to look into why this wasn't caught by the tests...

Comment (greschd) [reply aiida/orm/implementation/general/calculation/job/__init__.py]: Tests didn't catch this because the test which compares actual and stored hashes has a non-zero ``retrieve_temporary_list``. I guess that's ok, we can't test all possible combinations of inputs anyway.

On the other hand, it might be worth investigating why it's possible that an attribute is updatable, without being on the ``updatable_attributes`` list.

**Comment by giovannipizzi** [comment]
I'm not sure where @greschd wrote the comment on all attributes being updatable now... anyway, it's a big bug that was undiscovered and I opened #1109 for that.

**Comment by greschd** [comment]
Cool, thanks for merging!

For the ``_aiida_hash`` and ``_aiida_cached_from`` extras, I think we discussed at some point adding a column for the hash, and a link type for cached_from. I don't know how much effort both of these changes are, though. Probably we will want to use "internal" extras again in the future when developing new features, so maybe adding the logic to hide the ``_aiida_*`` extras still makes sense?
